File: rust-core/zx-server/src/screen_stream.rs
//! Screen-frame streaming TCP server, ported from
//! `zxspectrum/server/screen_stream.py`: on each connection, loops forever
//! pushing the live display as a length-prefixed PNG frame (4-byte
//! big-endian length + PNG bytes) until the client disconnects. Reads the
//! latest rendered screen via `Engine::watch_screen()` (bypasses the
//! command queue, same reason Python's version reads `machine` directly --
//! streaming shouldn't stall behind a live `run()`).

use image::{ImageBuffer, ImageEncoder, Rgb};
use std::io;
use std::time::Duration;
use tokio::io::AsyncWriteExt;
use tokio::net::{TcpListener, TcpStream};
use tokio::time::interval;
use zx_core::ula::{FULL_SCREEN_HEIGHT, FULL_SCREEN_WIDTH};
use zx_engine::Engine;

const FRAME_INTERVAL: Duration = Duration::from_millis(20); // 10fps

pub async fn serve(engine: Engine, host: &str, port: u16) -> io::Result<()> {
    let listener = TcpListener::bind((host, port)).await?;
    println!("Screen stream server listening on {host}:{port}");
    loop {
        let (stream, _) = listener.accept().await?;
        let engine = engine.clone();
        tokio::spawn(async move {
            let _ = handle_connection(stream, engine).await;
        });
    }
}

async fn handle_connection(mut stream: TcpStream, engine: Engine) -> io::Result<()> {
    let mut screen_rx = engine.watch_screen();
    let mut ticker = interval(FRAME_INTERVAL);
    loop {
        ticker.tick().await;
        let rgb = screen_rx.borrow_and_update().clone();
        if rgb.len() != FULL_SCREEN_WIDTH * FULL_SCREEN_HEIGHT * 3 {
            continue; // not primed yet
        }
        let png = encode_png(&rgb);
        let len = (png.len() as u32).to_be_bytes();
        if stream.write_all(&len).await.is_err() {
            break;
        }
        if stream.write_all(&png).await.is_err() {
            break;
        }
    }
    Ok(())
}

pub(crate) fn encode_png(rgb: &[u8]) -> Vec<u8> {
    let image: ImageBuffer<Rgb<u8>, _> =
        ImageBuffer::from_raw(FULL_SCREEN_WIDTH as u32, FULL_SCREEN_HEIGHT as u32, rgb.to_vec())
            .expect("screen buffer is the right size");
    let mut out = Vec::new();
    image::codecs::png::PngEncoder::new(&mut out)
        .write_image(
            &image,
            FULL_SCREEN_WIDTH as u32,
            FULL_SCREEN_HEIGHT as u32,
            image::ExtendedColorType::Rgb8,
        )
        .expect("PNG encoding cannot fail for a valid RGB buffer");
    out
}

File: cpp-core/src/screen_stream.cpp
// Screen-frame streaming TCP server, ported from
// rust-core/zx-server/src/screen_stream.rs: on each connection, loops forever
// pushing the live display as a length-prefixed PNG frame (4-byte big-endian
// length + PNG bytes) until the client disconnects.
//
// Reads the screen via `Engine::screen()`, which bypasses the command queue
// for exactly this reason -- streaming must not stall behind a live `run()`,
// which is precisely when there is something worth watching.

#include "screen_stream.h"

#include "net.h"
#include "ula.h"

#define STB_IMAGE_WRITE_IMPLEMENTATION
#define STBI_WRITE_NO_STDIO
#include <stb_image_write.h>

#include <chrono>
#include <cstdio>
#include <thread>

namespace zx {
namespace {

constexpr auto FRAME_INTERVAL = std::chrono::milliseconds(20);

void collect_png(void* context, void* data, int size) {
    auto* out = static_cast<std::vector<uint8_t>*>(context);
    const auto* bytes = static_cast<const uint8_t*>(data);
    out->insert(out->end(), bytes, bytes + size);
}

void handle_connection(net::Socket sock, Engine& engine) {
    constexpr size_t EXPECTED = size_t(FULL_WIDTH) * FULL_HEIGHT * 3;
    for (;;) {
        std::this_thread::sleep_for(FRAME_INTERVAL);
        const std::vector<uint8_t> rgb = engine.screen();
        if (rgb.size() != EXPECTED) {
            continue; // not primed yet
        }
        const std::vector<uint8_t> png = encode_png(rgb);
        const uint32_t len = uint32_t(png.size());
        const uint8_t header[4] = {uint8_t(len >> 24), uint8_t(len >> 16), uint8_t(len >> 8),
                                   uint8_t(len)};
        if (!sock.send_all(reinterpret_cast<const char*>(header), sizeof header)) {
            break;
        }
        if (!sock.send_all(reinterpret_cast<const char*>(png.data()), png.size())) {
            break;
        }
    }
}

} // namespace

std::vector<uint8_t> encode_png(const std::vector<uint8_t>& rgb) {
    std::vector<uint8_t> out;
    stbi_write_png_to_func(collect_png, &out, int(FULL_WIDTH), int(FULL_HEIGHT), 3, rgb.data(),
                           int(FULL_WIDTH) * 3);
    return out;
}

void serve_screen_stream(Engine& engine, const std::string& host, uint16_t port) {
    net::Listener listener;
    std::string error;
    if (!listener.listen(host, port, error)) {
        std::fprintf(stderr, "Screen stream server failed to start: %s\n", error.c_str());
        return;
    }
    std::printf("Screen stream server listening on %s:%u\n", host.c_str(), unsigned(port));
    std::fflush(stdout);

    for (;;) {
        net::Socket sock = listener.accept();
        if (!sock.valid()) {
            continue;
        }
        std::thread([s = std::move(sock), &engine]() mutable { handle_connection(std::move(s), engine); })
            .detach();
    }
}

} // namespace zx
